From e0a8fc169297bfe74615ac4f2422e5011099fe1b Mon Sep 17 00:00:00 2001
From: Alex Antony <alexx.kanjirathingal.antony@intel.com>
Date: Wed, 30 Nov 2022 17:18:04 +0000
Subject: [PATCH] Add Model to ProcessorSummary

In Redfish ComputerSystem schema, the ProcessorSummary parameter
lists summary information of the Processors on the system. This commit
adds the 'Model' property to ProcessorSummary.

If the CPU Models are different, then the 'Model' field takes the first
entry in alphabetical order.

Signed-off-by: Ali Ahmed <ama213000@gmail.com>
Signed-off-by: AppaRao Puli <apparao.puli@linux.intel.com>
Signed-off-by: Alex Antony <alexx.kanjirathingal.antony@intel.com>
---
 redfish-core/lib/systems.hpp | 35 ++++++++++++++++++++++++++++++++++-
 1 file changed, 34 insertions(+), 1 deletion(-)

diff --git a/redfish-core/lib/systems.hpp b/redfish-core/lib/systems.hpp
index caf6a292..07d243bb 100644
--- a/redfish-core/lib/systems.hpp
+++ b/redfish-core/lib/systems.hpp
@@ -134,8 +134,41 @@ inline void getProcessorProperties(
 
     BMCWEB_LOG_DEBUG << "Got " << properties.size() << " Cpu properties.";
 
-    // TODO: Get Model
+    const std::string* modelStr = nullptr;
 
+    const bool modelsuccess = sdbusplus::unpackPropertiesNoThrow(
+        dbus_utils::UnpackErrorPrinter(), properties, "Family", modelStr);
+
+    if (!modelsuccess)
+    {
+        return;
+    }
+
+    if ((modelStr != nullptr) && (*modelStr !=""))
+    {
+        nlohmann::json& prevModel =
+            aResp->res.jsonValue["ProcessorSummary"]["Model"];
+        std::string* prevModelPtr = prevModel.get_ptr<std::string*>();
+
+        // If CPU Models are different, use the first entry in
+        // alphabetical order
+
+        // If Model has never been set
+        // before, set it to *modelStr
+        if (prevModelPtr == nullptr)
+        {
+            prevModel = *modelStr;
+        }
+        // If Model has been set before, only change if new Model is
+        // higher in alphabetical order
+        else
+        {
+            if (*modelStr < *prevModelPtr)
+            {
+                prevModel = *modelStr;
+            }
+        }
+    }
     const uint16_t* coreCount = nullptr;
 
     const bool success = sdbusplus::unpackPropertiesNoThrow(
-- 
2.25.1

